### ISSUE TITLE
Google Cloud GCP Life Sciences API - Shared VPC support, Service Account, Creating PD and Support NFS

### ISSUE THREAD

**Issue by thomasnyc** [snakemake/snakemake#1563]
I learn about Snakemake and Google Cloud (GCP) Life Sciences API and there are 4 different features are very useful to GCP users. Here are my suggestions: 

1. Shared VPC support - in most large accounts / infrastructure setup. Shared VPC is a very common network setup. It is important to add support from the Snakemake submission to specific which network to be used. I believe currently Snakemake instruct LSAPI to leverage default network only. 
eg Flags like : --host_gcp_project_name <host project id> --subnetwork_name <subnetwork name>

2. Specific Service Account - in order to apply different IAM permissions for different users / workflow. Preferably specifying  different service accounts  when submitting the Snakemake job to LSAPI . Currently, LSAPI only uses the default compute service account which is NOT sufficient in a multiple user environment. 
eg Flag:  --service_account <specific service account> --service_account_json <json key location and name>

3. Persistent Disk support - From genomic workload, users love to bring data to the LSAPI worker nodes. One of the effective ways to do this at scale is via PD creation with different PD images. Allowing the LSAPI worker creating different size of PD with specific images is needed via Snakemake. Also, it is a good idea to have a fast and large temp working space for input and output data. 
eg Flag: --pd_size <# of GB> --pd_type <SSD, standard ,etc> -- image_project <gcp project hosting image> --pd_image <image name> 

4. NFS support - Again, bring data to LSAPI nodes can leverage NFS share. It would be ideal to allow NFS mount at the LSAPI host. This also allows users to monitor the job status and output files real time. 
eg Flag: --nfs_mount <10.x.x.x/nfssharename >

**Comment by cademirch** [comment]
I have thought about implementing 3, as it would be nice to have a choice of disk type depending on the workload. However I'm a bit stuck as currently the disk for the VM is set here: https://github.com/snakemake/snakemake/blob/b0b206c4c34616093d7655789487f923b51e365b/snakemake/executors/google_lifesciences.py#L555-L560

The [GLS docs](https://cloud.google.com/life-sciences/docs/reference/rpc/google.cloud.lifesciences.v2beta#virtualmachine) shows that the virtual machine can take a list of disks to attach, though its not clear to me how you would make sure the actions happen on a specified disk. Perhaps @vsoch can help?

**Comment by vsoch** [comment]
For 3, if snakemake can easily use Google Storage why would we need to attach extra disks?

**Comment by cademirch** [comment]
Hmm I'm not sure I follow. I think the idea is to have the option to specify the disk type for a rule. So for heavy I/O workloads you could have a VM with pd-SSD.

**Comment by vsoch** [comment]
Oh I see, I read:

> From genomic workload, users love to bring data to the LSAPI worker nodes.

As "I need a way to save and bring data to the nodes" and I think I'd use storage for that. If there is a different use case to just further customize the VM configuration, I don't see any issue with trying that out.

**Comment by thomasnyc** [comment]
Thank you for looking into this.  Let me elaborate on the point 3.
Data can be bring to the LSAPI node multiple ways. Of course, GCS copy is the most common way. 
For large number of instance (100s) and large dataset (1TB - 5TB) scenario , Copy from GCS to each instance takes longer time. If we use an PD image creating separate PD per LSAPI instance, The instance with dataset available only takes  ~5 mins. 
Also, with PD, we can easily setup for heavy IO workload , temp space and enough space for datasets being copy to the instance.

**Comment by cademirch** [comment]
Ah so if I'm understanding correctly, @thomasnyc's suggestion is some where between me and @vsoch's interpretation. I'm still not entirely familiar with GCP (and _the _cloud__) in general, so I'm not sure how one would initialize a PD with a dataset. However, given a solution to that - I think I could see this being doable. 

As for my suggestion - which is to have a new PD created of a user-specified type (e.g SSD) for the VM. Snakemake would still download the input files, but instead of to the boot disk, it would be on this PD. That way for heavy IO, one could take advantage of faster SSDs. My current hangup in implementing this is that I don't understand how to ensure that the Snakemake container uses the PD and not the boot disk.

**Comment by vsoch** [comment]
Yep I understand now - he wants to have a disk that exists in advance that can provide data. If you need to share between workers it would need to be mounted read only.

**Comment by qinqian** [comment]
Thanks @thomasnyc for requesting the features. I think the feature 2 is very useful, is there any update for point 2 Specific Service Account?
